{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "75d7ebc0-66b0-4602-87fe-1dc79999f928",
   "metadata": {},
   "source": [
    "## Instructions for Performing Inference on HEST-Bench with STPath\n",
    "\n",
    "This tutorial will guide you to:\n",
    "* Prepare the data (AnnData object and visual embeddings) from a selected dataset in HEST-Bench for inference.\n",
    "* Run inference using STPath and compute the corresponding evaluation metrics.\n",
    "\n",
    "It is important to note that there are differences between the data in [HEST-1K](https://huggingface.co/datasets/MahmoodLab/hest/tree/main) and [HEST-Bench](https://huggingface.co/datasets/MahmoodLab/hest-bench/tree/main). Specifically, HEST-Bench is a subset of HEST-1K. The tissue segmentation in HEST-1K has been refined in its latest release (see this [issue](https://github.com/mahmoodlab/HEST/issues/68) for details). In our experiments in the paper, we use the same dataset IDs and highly variable genes (HVGs) defined in HEST-Bench, but the actual samples are taken from HEST-1K. For simplicity, however, this tutorial will demonstrate the workflow using the data provided directly by HEST-Bench."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "582bbfd7-65ac-43eb-a735-37e3f2090dec",
   "metadata": {},
   "source": [
    "### Download dataset from HuggingFace\n",
    "\n",
    "We here specifically download the LUNG dataset from HEST-Bench."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4dd9239f-ba38-4546-bdd0-735428027e4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9286ff30-2f9a-4d96-818e-3d55ae1fc765",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "df6a465eda534b6f872e84b1272ccab7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Fetching 11 files:   0%|          | 0/11 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "20ab1847f6e04a318de9e0f08b5bb6d6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX118.png:   0%|          | 0.00/1.43M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "83c398f49db94474a9600e86deec40db",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX141.png:   0%|          | 0.00/2.40M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b3db9dfce36b4407867d82cedb3d3ab0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX118.h5:   0%|          | 0.00/301M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "61891c753c274f94b8df0240fe4ca721",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX141.h5ad:   0%|          | 0.00/6.84M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6a4417eed0784d89a952d204082cd8c2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX141.h5:   0%|          | 0.00/505M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "df057ce3c99b40a5a473f5c60f576367",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "test_0.csv:   0%|          | 0.00/79.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a8e4270937fb469e830be4cfaf49e163",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "TENX118.h5ad:   0%|          | 0.00/4.80M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "94a0aaa8035846f4b6d8f6e9a930436e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "test_1.csv:   0%|          | 0.00/79.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "438a8ef69bf648a286ef52d93f728aa8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "train_0.csv:   0%|          | 0.00/79.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a810afed140c4fb48fc55e0a2c14293f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "train_1.csv:   0%|          | 0.00/79.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5819ab4ecb784aea9fd7ad92eb98db4f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "var_50genes.json:   0%|          | 0.00/452 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'/scratch/ti.huang/hest-bench'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from huggingface_hub import snapshot_download\n",
    "\n",
    "snapshot_download(\n",
    "    repo_id=\"MahmoodLab/hest-bench\", \n",
    "    repo_type='dataset', \n",
    "    local_dir='/home/ti.huang/scratch/hest-bench',\n",
    "    allow_patterns=['LUNG/']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8493a992-77e3-468c-a234-869b65e98fa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['adata', 'patches', 'splits', 'var_50genes.json']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata includes the spatial transcriptomics profiles\n",
    "# patches includes the image patches\n",
    "# splits includes the multi-fold split, which is not used in our setup\n",
    "# var_50genes.json is the top 50 HVG list\n",
    "os.listdir('/home/ti.huang/scratch/hest-bench/LUNG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "42d50226-b454-4119-8311-d4c14ca029e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TENX118.h5ad', 'TENX141.h5ad']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# LUNG dataset includes two samples\n",
    "os.listdir('/home/ti.huang/scratch/hest-bench/LUNG/adata')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5113f99b-42f5-4d71-97fc-02d5fca2f44f",
   "metadata": {},
   "source": [
    "### Initialze the Gigapath encoder\n",
    "\n",
    "Please first download the pretrained weight of Gigapath (`https://huggingface.co/prov-gigapath/prov-gigapath/blob/main/pytorch_model.bin`) in `your_weight_dir/gigapath`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5969072-57ea-44ca-936d-a2f7ec6bfc83",
   "metadata": {},
   "outputs": [],
   "source": [
    "from stpath.app.preprocess.hest import LazyEncoder\n",
    "\n",
    "# weight_root should be `your_weight_dir`\n",
    "lazy_enc = LazyEncoder(\"gigapath\", weights_root=\"/home/ti.huang/project/single_cell_dataset/weights_root\")\n",
    "gigapath_encoder, img_transforms = lazy_enc.get_model(device=0)  # set the gpu device as 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9631728-7f91-46a6-9ba8-991b151e5dca",
   "metadata": {},
   "source": [
    "Encode each image patch with Gigapath and save them in `embeddings` folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "77fae6df-8999-4470-9a0f-5d9f9d0ff266",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Embedding Tiles TENX118: 100%|██████████████████████████████████████| 16/16 [00:37<00:00,  2.32s/it]\n",
      "Embedding Tiles TENX141: 100%|██████████████████████████████████████| 26/26 [01:02<00:00,  2.41s/it]\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from stpath.app.preprocess.hest import embed_tiles\n",
    "from stpath.hest_utils.st_dataset import H5TileDataset\n",
    "\n",
    "batch_size = 128\n",
    "\n",
    "for sample_id in ['TENX118', 'TENX141']:\n",
    "    tile_h5_path = os.path.join('/home/ti.huang/scratch/hest-bench/LUNG', \"patches\", f\"{sample_id}.h5\")\n",
    "    assert os.path.isfile(tile_h5_path), f\"Tile file {tile_h5_path} does not exist\"\n",
    "    embedding_dir = os.path.join('/home/ti.huang/scratch/hest-bench/LUNG/embeddings', sample_id, lazy_enc.name, 'fp32')\n",
    "    os.makedirs(embedding_dir, exist_ok=True)\n",
    "    embed_path = os.path.join(embedding_dir, f'{sample_id}.h5')\n",
    "    \n",
    "    tile_dataset = H5TileDataset(tile_h5_path, chunk_size=batch_size, img_transform=img_transforms)\n",
    "    tile_dataloader = torch.utils.data.DataLoader(\n",
    "                       tile_dataset, \n",
    "                       batch_size=1, \n",
    "                       shuffle=False,\n",
    "                       num_workers=1\n",
    "                    )\n",
    "    _ = embed_tiles(sample_id, tile_dataloader, encoder, embed_path, device=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4080035c-27b5-4e12-9956-1a04f1f3e3e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TENX118', 'TENX141']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir('/home/ti.huang/scratch/hest-bench/LUNG/embeddings')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a7a37d3-6faa-4ecb-8ba5-3ad1b15f0ba8",
   "metadata": {},
   "source": [
    "### Perform inference with STPath"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6551e09b-97b9-428b-9a47-d2b4ba400e78",
   "metadata": {},
   "source": [
    "Initialize the STPath inference agent using the pretrained weights downloaded from `https://huggingface.co/tlhuang/STPath/tree/main`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "74cd9380-7bb5-4ec5-9237-5ea870edc98e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_genes: 38984, n_tech: 5, n_species: 6, n_organs: 25, n_cancer_annos: 5, n_domain_annos: 10\n",
      "Model loaded from /home/ti.huang/project/stfm/stpath/backup/stfm.pth\n"
     ]
    }
   ],
   "source": [
    "from stpath.app.pipeline.inference import STPathInference\n",
    "\n",
    "agent = STPathInference(\n",
    "    gene_voc_path='/home/ti.huang/STPath/utils_data/symbol2ensembl.json',\n",
    "    model_weight_path='/home/ti.huang/project/stfm/stpath/backup/stfm.pth', \n",
    "    device=0\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d7b343-9d6d-45b1-b1a4-73ebd5d0257f",
   "metadata": {},
   "source": [
    "Perform inference and save the predicted expressions of the HVGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "bbf06641-aaad-4bf6-908a-23f8b1d450cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting inference...\n",
      "Return results...\n",
      "Starting inference...\n",
      "Return results...\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "from stpath.hest_utils.st_dataset import load_adata\n",
    "from stpath.hest_utils.file_utils import read_assets_from_h5\n",
    "\n",
    "all_pred, all_gt = [], []\n",
    "for sample_id in ['TENX118', 'TENX141']:\n",
    "    source_dataroot = '/home/ti.huang/scratch/hest-bench/LUNG/'  # the root directory of the STPath repository\n",
    "    with open(os.path.join(source_dataroot, \"var_50genes.json\")) as f:\n",
    "        hvg_list = json.load(f)['genes']\n",
    "\n",
    "    data_dict, _ = read_assets_from_h5(os.path.join(source_dataroot, f\"embeddings/{sample_id}/gigapath/fp32/{sample_id}.h5\"))  # load the data from the h5 file\n",
    "    coords = data_dict[\"coords\"]\n",
    "    embeddings = data_dict[\"embeddings\"]\n",
    "    barcodes = data_dict[\"barcodes\"].flatten().astype(str).tolist()\n",
    "    adata = ad.read_h5ad(os.path.join(source_dataroot, f\"adata/{sample_id}.h5ad\"))[barcodes, :]\n",
    "    \n",
    "    # The return pred_adata includes the expressions of the genes in hvg_list, which is a list of highly variable genes.\n",
    "    pred_adata = agent.inference(\n",
    "        coords=coords, \n",
    "        img_features=embeddings, \n",
    "        organ_type=\"Lung\", \n",
    "        tech_type=None,\n",
    "        save_gene_names=hvg_list  # we only need the highly variable genes for evaluation\n",
    "    )\n",
    "\n",
    "    gt = np.log1p(adata[:, hvg_list].X.toarray())  # sparse -> dense\n",
    "\n",
    "    all_gt.append(gt)\n",
    "    all_pred.append(pred_adata.X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d59ec50-8ae6-4490-8e0a-d2cf27c66078",
   "metadata": {},
   "source": [
    "Concatenate the expressions of HVGs across the samples (TENX118, TENX141) and compute the average Pearson correlation. Note that the resulting metric (0.607) is slightly higher than the value reported in the paper (0.559 in Figure 2(c)), due to differences between the samples in HEST-1K and HEST-Bench."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "853efa2c-b28e-4179-977c-a23a105b66a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson correlation for LUNG dataset: 0.6075147917866707\n"
     ]
    }
   ],
   "source": [
    "all_gt = np.concatenate(all_gt, axis=0)\n",
    "all_pred = np.concatenate(all_pred, axis=0)\n",
    "# calculate the Pearson correlation coefficient between the predicted and ground truth gene expression\n",
    "all_pearson_list = []\n",
    "# go through each gene in the highly variable genes list\n",
    "for i in range(len(hvg_list)):\n",
    "    pearson_corr, _ = pearsonr(all_gt[:, i], all_pred[:, i])\n",
    "    all_pearson_list.append(pearson_corr.item())\n",
    "print(f\"Pearson correlation for LUNG dataset: {np.mean(all_pearson_list)}\")  # 0.1562"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
